combinatorics_and_monte_carlo/pi_functions.py

```diff
@@ -0,0 +1,58 @@
+"""Estimate pi using a Monte Carlo Method
+
+Simulates rain drops falling onto a square with an inscribed circle drawn in
+the sand. The proportion of raindrops that fall into the circle as compared to
+just the square approximates pi/4.
+
+Different implementations illustrate the advantages and disadvantages of
+using numpy.
+"""
+
+
+import random
+import math
+
+import numpy as np
+
+
+def std_pi(drops):
+    drops_in_circle_segment = 0
+
+    # simulate *drops* rain drops inside a square of a unit area and
+    # count how many of them fall inside an inscribed circle segment
+    for i in range(drops):
+        drop_distance_from_origin = math.hypot(
+            random.random(), random.random()
+            )
+        if drop_distance_from_origin < 1:
+            drops_in_circle_segment += 1
+    # ratio of drops in circle segment to total drops is ~ pi/4
+    return 4*drops_in_circle_segment/drops
+
+
+def std_pi_compact(drops):
+    # same as std_pi, but expressed in most compact form
+    return 4 * sum(
```

Comment:
The following thing inside `sum()` is a *generator expression*.
The only syntactical difference to a list comprehension is that there are no square brackets around it, but
regular parentheses (`()`, which are often optional and have been omitted here for brevity.
The big difference in behaviour is that the elements of a generator expression are generated on demand
(here: when `sum()` asks for the next value to add) and never stored as a whole (the same behaviour as a `range()`).